FreeRTOS/Demo/WIN32-MingW/main_blinky.c

```diff
@@ -182,84 +186,124 @@ TickType_t xNextWakeTime;
-		/* To get here something must have been received from the queue, but
-		is it an expected value?  Normally calling printf() from a task is not
-		a good idea.  Here there is lots of stack space and only one task is
-		using console IO so it is ok.  However, note the comments at the top of
-		this file about the risks of making Windows system calls (such as
-		console output) from a FreeRTOS task. */
-		if( ulReceivedValue == mainVALUE_SENT_FROM_TASK )
-		{
-			printf( "Message received from task\r\n" );
-		}
-		else if( ulReceivedValue == mainVALUE_SENT_FROM_TIMER )
-		{
-			printf( "Message received from software timer\r\n" );
-		}
-		else
-		{
-			printf( "Unexpected message\r\n" );
-		}
-
-		fflush( stdout );
-	}
+    /* Prevent the compiler warning about the unused parameter. */
+    ( void ) pvParameters;
+
+    for( ;; )
+    {
+        /* Wait until something arrives in the queue - this task will block
+        indefinitely provided INCLUDE_vTaskSuspend is set to 1 in
+        FreeRTOSConfig.h.  It will not use any CPU time while it is in the
+        Blocked state. */
+        xQueueReceive( xQueue, &ulReceivedValue, portMAX_DELAY );
+
+        /* Enter critical section to use printf. Not doing this could potentially cause
+        a deadlock if the FreeRTOS simulator switches contexts and another task
+        tries to call printf - it should be noted that use of printf within
+        the FreeRTOS simulator is unsafe, but used here for simplicity. */
+        taskENTER_CRITICAL();
+        {
+            /* To get here something must have been received from the queue, but
+            is it an expected value?  Normally calling printf() from a task is not
+            a good idea.  Here there is lots of stack space and only one task is
+            using console IO so it is ok.  However, note the comments at the top of
+            this file about the risks of making Windows system calls (such as
+            console output) from a FreeRTOS task. */
+            if( ulReceivedValue == mainVALUE_SENT_FROM_TASK )
+            {
+                printf( "Message received from task\r\n" );
+            }
+            else if( ulReceivedValue == mainVALUE_SENT_FROM_TIMER )
+            {
+                printf( "Message received from software timer\r\n" );
+            }
+            else
+            {
+                printf( "Unexpected message\r\n" );
+            }
+
+            fflush( stdout );
+        }
+        taskEXIT_CRITICAL();
+    }
 }
+
 /*-----------------------------------------------------------*/
 
+/* Called from prvKeyboardInterruptHandler(), which is defined in main.c. */
+void vBlinkyKeyboardInterruptHandler( int xKeyPressed )
+{
+    /* Handle keyboard input. */
+    switch ( xKeyPressed )
+    {
+
+    case mainRESET_TIMER_KEY:
+
+        if ( xTimer != NULL )
+        {
+            /* Critical section around printf to prevent a deadlock
+               on context switch. */
+            portENTER_CRITICAL();

```

Comment:
Was there a reason for using the portENTER/portEXIT macros over the taskENTER/EXIT ones? Asking since they're used interchangeably in this demo but in the other ones it was only the taskENTER/EXIT ones.

Reply:
For winsim they technically do the same thing, I used portENTER_CRITICAL here because this is not a task but an interrupt handler.

Reply:
Gotcha, in that case do you want to swap over the task calls to these port ones for the sake of consistency?

Reply:
I use the task calls in tasks as you would normally.